### ISSUE TITLE
IPv6 address with :0: address group causes parsing to fail

### ISSUE THREAD

**Issue by mdeshon** [secdev/scapy#502]
I encountered an error when doing an 'import scapy.all' which was more specifically triggered by scapy.arch.linux.read_route6(). The origin of the issue is when one of the host interfaces has an IPv6 allocation where the network address contains a zero group (e.g. 2620:0:1000::/40), the 0 is normalized to ::, then later when finding the subnet the string handling would try to operate on 2000::1000:: which is an illegal encoding since the location of zero groups is indeterminate.

I was able to eliminate the error for my purposes through a simple change:

In the file
/usr/local/lib/python2.7/dist-packages/scapy/arch/linux.py

```
--- a/linux.py	2017-02-06 12:11:18.791547717 -0800
+++ b/linux.py	2017-02-06 12:12:09.160076529 -0800
@@ -257,7 +257,7 @@
     def proc2r(p):
         ret = struct.unpack('4s4s4s4s4s4s4s4s', p)
         ret = ':'.join(ret)
-        return scapy.utils6.in6_ptop(ret)
+        return ret
     
     lifaddr = in6_getifaddr() 
     for l in f.readlines():
```

in other words, returning 'ret' from proc2r() rather than first normalizing it.

**Comment by guedou** [comment]
Thanks for reporting this.  Could you share the Scapy version you are using, the Linux distribution and the content of   /proc/net/ipv6_route ?

That will help us reproduce the issue and add a relevant unit test to trigger it.

**Comment by mdeshon** [comment]
Output of cat /proc/net/ipv6_route attached.

$ pip show scapy
---
Name: scapy
Version: 2.3.3
Location: /usr/local/lib/python2.7/dist-packages
Requires:

I'm running a corporate-modified form of Ubuntu.

I don't want to dump my whole routing table in a public forum, but I
believe this line will trigger the issue:

26200000100021100000000000000000 40 00000000000000000000000000000000 00
00000000 000000000000000000000000 00000100 00000000 00000000 004c0001
em1


On Mon, Feb 6, 2017 at 10:43 PM, Guillaume Valadon <notifications@github.com
> wrote:

> Thanks for reporting this. Could you share the Scapy version you are
> using, the Linux distribution and the content of /proc/net/ipv6_route ?
>
> That will help us reproduce the issue and add a relevant unit test to
> trigger it.
>
>
> Sent from my iPhone
>

**Comment by guedou** [comment]
Thanks for the information.

Unfortunately, I did not manage to reproduce the issue with the following mocked script. Could it be related to libc instead ?


```python
from scapy.all import *
import mock

data = "26200000100021100000000000000000 40 00000000000000000000000000000000 00 00000000000000000000000000000000 00000100 00000000 00000000 004c0001 scapy0\n"

@mock.patch("scapy.arch.linux.in6_getifaddr")
@mock.patch("scapy.arch.linux.open", mock.mock_open(read_data=data))
def test_read_routes6(mock_in6_getifaddr):
    mock_in6_getifaddr.return_value = [("2620:0:1000:2110::1", 0, "scapy0")]
    return read_routes6()

print test_read_routes6()
```

```shell
python test_route6.py
[('2620:0:1000:2110::', 64, '::', 'scapy0', ['2620:0:1000:2110::1'])]
```

**Comment by guedou** [comment]
Issue closed due to no activity.

**Comment by andysouthwell2** [comment]
Hi,

Seeing the same issue here with python 2.7.2.5 on Windows 7 with scapy 2.3.3. The following reproduces the issue for me, in the interactive python console (just typing scapy at the command-line) :-
>>>a=IPv6(src='fe80::200:0:0:1',dst='fe80::d4ca:6eff:fe77:6656')
>>>ls(a)
(shows me that src has the value 'fe80::200::1', which is illegal)
and then
>>>a.show()
generates an exception in pton_ntop.py line 40, "Illegal syntax for IP address")

**Comment by andysouthwell2** [comment]
I have now confirmed the bug is in your inet_ntop(), but I don't know where yet. This reproduces the issue:

a=inet_pton(socket.AF_INET6, 'fe80::ff:0:0:1')
print(inet_ntop(socket.AF_INET6, a))

**Comment by andysouthwell2** [comment]
The problem is the way that inet_ntop strips trailing zeros and reduces array elements down into groups of colons. Unfortunately it fails to ensure that it doesn't end up with multiple double-colon occurrences. The best solution would be to determine the position and length of the longest range of zeros beforehand, and only replace those.

**Comment by andysouthwell2** [comment]
New ticket 978 opened for this issue (given this one is now closed), and patch provided.
